clients/src/main/java/org/apache/kafka/common/network/BanzaiTransportLayer.java

```diff
@@ -0,0 +1,192 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.network;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.SelectionKey;
+
+/**
+ * Read buffered {@link PlaintextTransportLayer} to be able to push back on receiving buffer
+ */
+public class BanzaiTransportLayer extends PlaintextTransportLayer {
+    private static final Logger log = LoggerFactory.getLogger(BanzaiTransportLayer.class);
+    private static final int BUFF = 65536;
+    private ByteBuffer netReadBuffer;
+    private boolean authenticationDone;
+
+    public BanzaiTransportLayer(SelectionKey key) throws IOException {
+        super(key);
+        netReadBuffer = ByteBuffer.allocate(BUFF);
+        authenticationDone = false;
+    }
+
+    /**
+     * No buffer shadowing needed
+     */
+    public void authenticationDone() {
+        authenticationDone = true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public long read(ByteBuffer[] dsts, int offset, int length) throws IOException {
+        if ((offset < 0) || (length < 0) || (offset > dsts.length - length))
+            throw new IndexOutOfBoundsException();
+
+        int totalRead = 0;
+        int i = offset;
+        while (i < length) {
+            if (dsts[i].hasRemaining()) {
+                int read = read(dsts[i]);
+                if (read > 0)
+                    totalRead += read;
+                else
+                    break;
+            }
+            if (!dsts[i].hasRemaining()) {
```

Comment:
You can just use an else branch here, so the condition does not get evaluated twice each loop